core/lib/dal/src/snapshots_dal.rs

```diff
@@ -0,0 +1,110 @@
+impl SnapshotsDal<'_, '_> {
+    pub async fn add_snapshot(
+        &mut self,
+        l1_batch_number: L1BatchNumber,
+        files: &[String],
+    ) -> Result<(), sqlx::Error> {
+        sqlx::query!(
+            "INSERT INTO snapshots (l1_batch_number, created_at, files) \
+             VALUES ($1, now(), $2)",
+            l1_batch_number.0 as i32,
+            files
+        )
+        .execute(self.storage.conn())
+        .await?;
+        Ok(())
+    }
+
+    pub async fn get_snapshots(&mut self) -> Result<AllSnapshots, sqlx::Error> {
+        let records: Vec<SnapshotBasicMetadata> =
+            sqlx::query!("SELECT l1_batch_number, created_at FROM snapshots")
+                .fetch_all(self.storage.conn())
+                .await?
+                .into_iter()
+                .map(|r| SnapshotBasicMetadata {
+                    l1_batch_number: L1BatchNumber(r.l1_batch_number as u32),
+                    generated_at: DateTime::<Utc>::from_naive_utc_and_offset(r.created_at, Utc),
+                })
+                .collect();
+
+        Ok(AllSnapshots { snapshots: records })
+    }
+
+    pub async fn get_snapshot(
+        &mut self,
+        l1_batch_number: L1BatchNumber,
+    ) -> Result<Option<SnapshotFullMetadata>, sqlx::Error> {
+        let record = sqlx::query!(
+            "SELECT l1_batch_number, created_at, files \
+            FROM snapshots WHERE l1_batch_number = $1",
+            l1_batch_number.0 as i32
+        )
+        .fetch_optional(self.storage.conn())
+        .await?;
+
+        Ok(record.map(|r| SnapshotFullMetadata {
+            l1_batch_number: L1BatchNumber(r.l1_batch_number as u32),
+            generated_at: DateTime::<Utc>::from_naive_utc_and_offset(r.created_at, Utc),
+            storage_logs_files: r.files,
+        }))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::ConnectionPool;
+    use db_test_macro::db_test;
+    use zksync_types::L1BatchNumber;
+
+    #[db_test(dal_crate)]
+    async fn adding_snapshot(pool: ConnectionPool) {
+        let mut conn = pool.access_storage().await.unwrap();
+        let mut dal = conn.snapshots_dal();
+        let l1_batch_number = L1BatchNumber(100);
+        dal.add_snapshot(l1_batch_number, &[])
+            .await
+            .expect("Failed to add snapshot");
+
+        let snapshot = dal
+            .get_snapshot(l1_batch_number)
+            .await
+            .expect("Failed to retrieve snapshot");
+        assert!(snapshot.is_some());
+        assert_eq!(
+            snapshot.unwrap().l1_batch_number,
+            l1_batch_number as L1BatchNumber
+        );
+    }
+
+    #[db_test(dal_crate)]
+    async fn adding_files(pool: ConnectionPool) {
+        let mut conn = pool.access_storage().await.unwrap();
+        let mut dal = conn.snapshots_dal();
+        let l1_batch_number = L1BatchNumber(100);
+        dal.add_snapshot(
+            l1_batch_number,
+            &["test_file1.bin".to_string(), "test_file2.bin".to_string()],
```

Comment:
Maybe a stupid question: What are files supposed to be on the DB level and on the API level? Are they the same? (I guess a key in a publicly accessible GCS bucket – if so, is it assumed that the GCS bucket URL is known by clients?)

Reply:
>  if so, is it assumed that the GCS bucket URL is known by clients?

Interesting - in other places where we use GCS we store file names (just like @tomg10 did here). But this one is different as it deals with publicly accessible files. I was thinking we'd store file names (or even full urls) here and will concatenate them with the url base in the API and return such links to the users.